### PR TITLE
feat(cli): add `xds upgrade` command with v0.0.2 codemods

### LINKED ISSUE #484: feat(cli): version-to-version codemods via `npx xds upgrade`
## Summary

Ship version-to-version codemods inside `@xds/cli` so consumers can migrate breaking API changes automatically. The upgrade path becomes:

```bash
npx xds upgrade          # dry-run: show what would change
npx xds upgrade --apply  # apply the transforms
```

LLMs do the same thing: bump the version in package.json, run `npx xds upgrade --apply`, done. This lets us move fast and make breaking changes without burdening consumers.

## Motivation

XDS is iterating quickly — API unification (prop renames, callback standardization, naming consistency) means breaking changes between versions. In an AI-native world, the migration story needs to be:

1. Read the changelog (auto-generated from changesets)
2. Run one command
3. Done

String-based find/replace is unsafe — common prop names like `items` or `tooltip` appear everywhere. We need **AST-level transforms** (via jscodeshift) scoped to specific XDS components and props.

### File structure

```
packages/cli/src/
├── commands/
│   └── upgrade.mjs              # CLI command
└── codemods/
    ├── registry.mjs              # Version → transforms mapping
    ├── runner.mjs                # Orchestration: detect version, resolve, run
    ├── ensure-jscodeshift.mjs    # Lazy-install jscodeshift on first use
    └── transforms/
        └── v0.0.2/
            ├── index.mjs         # Manifest: description + transforms list
            ├── rename-selector-items-to-options.mjs
            ├── unify-visibility-to-onOpenChange.mjs
            ├── unify-uncontrolled-to-defaultX.mjs
            ├── rename-banner-endButton-to-endContent.mjs
            ├── rename-form-tooltip-startIcon.mjs
            └── __tests__/
                └── (one test per transform with fixture pairs)
```

### CLI UX

```bash
# Preview what would change (dry-run, default)
npx xds upgrade

# Apply changes
npx xds upgrade --apply

# Target a specific version
npx xds upgrade --to 0.0.3

# Run one specific codemod only
npx xds upgrade --codemod rename-selector-items-to-options

# Specify source directory (default: ./src)
npx xds upgrade --path ./app
```

### Flow

1. Read consumer's `package.json` → find current `@xds/core` version
2. Determine target version (latest known in registry, or `--to` flag)
3. Look up all transform sets between current and target
4. If jscodeshift not installed → prompt to install (`npm install --no-save jscodeshift`)
5. For each version in order:
   - Print version header + description
   - Run each transform (dry-run or apply)
   - Report: files scanned, files modified, changes made
6. Summary: total files changed, suggest running lint/typecheck

### Transform anatomy

Each transform is a standard jscodeshift transform with metadata:

```js
// transforms/v0.0.2/rename-selector-items-to-options.mjs
export default function transform(file, api) {
  const j = api.jscodeshift;
  const root = j(file.source);
  let hasChanges = false;

  // Find <XDSSelector items={...}> → <XDSSelector options={...}>
  root
    .find(j.JSXOpeningElement, { name: { name: 'XDSSelector' } })
    .find(j.JSXAttribute, { name: { name: 'items' } })
    .forEach(path => {
      path.node.name.name = 'options';
      hasChanges = true;
    });

  return hasChanges ? root.toSource() : undefined;
}

export const meta = {
  title: 'Rename Selector items → options',
  description: 'XDSSelector prop `items` was renamed to `options` for clarity.',
  pr: '#479',
};
```

### Version manifest

```js
// transforms/v0.0.2/index.mjs
export const version = '0.0.2';
export const description = 'API unification: consistent naming across components';
export const transforms = [
  './rename-selector-items-to-options.mjs',
  './unify-visibility-to-onOpenChange.mjs',
  './unify-uncontrolled-to-defaultX.mjs',
  './rename-banner-endButton-to-endContent.mjs',
  './rename-form-tooltip-startIcon.mjs',
];
```

## Design principles

1. **AST transforms only** — no string replacement. Precise, safe, scoped to specific components/props.
2. **Idempotent** — running the same codemod twice produces the same result.
3. **Dry-run by default** — preview changes before writing. `--apply` to commit.
4. **Version-sequenced** — jumping from 0.0.1 → 0.0.5 runs v0.0.2, v0.0.3, v0.0.4, v0.0.5 in order.
5. **Testable** — each transform has input/output fixture pairs.
6. **LLM-friendly** — transforms have metadata (title, description, PR link) that LLMs can read.
7. **Lazy jscodeshift** — not bundled with the CLI. Installed on first use to keep `@xds/cli` lean.

## Changeset integration

When a changeset describes a breaking change, reference the codemod:

```md
---
'@xds/core': minor
---

Renamed `items` prop to `options` on XDSSelector.

**Codemod:** `npx xds upgrade --codemod rename-selector-items-to-options`
```

The auto-generated CHANGELOG tells consumers exactly how to migrate. LLMs read the changelog → run the command → done.

## v0.0.2 transforms

| Transform | What it does | PR |
|---|---|---|
| `rename-selector-items-to-options` | `<XDSSelector items={}>` → `<XDSSelector options={}>` | #479 |
| `unify-visibility-to-onOpenChange` | `onShow`/`onHide`/`onToggle` → `onOpenChange` on visibility components | #473 |
| `unify-uncontrolled-to-defaultX` | `initialOpen`/`initialValue` → `defaultOpen`/`defaultValue` | #470 |
| `rename-banner-endButton-to-endContent` | `<XDSBanner endButton={}>` → `<XDSBanner endContent={}>` | #437 |
| `rename-form-tooltip-startIcon` | `tooltip` → `labelTooltip`, `startIcon` → `labelIcon` on form components | #375 |

## Future considerations

- **Tree-shaking codemods** — only download transforms for versions you're actually upgrading through
- **CSS codemods** — if we ship pre-compiled CSS (#450), token/class renames could be handled too
- **Reverse codemods** — downgrade path for rollbacks
- **CI integration** — `npx xds upgrade --check` as a CI step that fails if codemods haven't been applied

### PR DESCRIPTION
**Pull request by cixzhang**
## What

Adds an `xds upgrade` CLI command that runs version-to-version codemods using jscodeshift to automate API migrations. Also adds the upgrade command to the agent docs injection so LLMs know to run codemods after version bumps.

## CLI Usage

```bash
# Dry run (default) — shows what would change
npx xds upgrade

# Apply changes
npx xds upgrade --apply

# Target a specific version
npx xds upgrade --to 0.0.2

# Run a single codemod
npx xds upgrade --codemod rename-selector-items-to-options

# Scan a custom directory
npx xds upgrade --path ./app

# List available codemods
npx xds upgrade --list
```

## How it works

1. Detects the consumer's `@xds/core` version from their `package.json`
2. Resolves which codemods need to run between current → target version
3. Lazy-installs `jscodeshift` on first use (keeps the CLI lean — no heavy deps by default)
4. Runs transforms against source files, reporting changes
5. In dry-run mode (default), shows what would change without writing files

## v0.0.2 Transforms

| Codemod | What it does | PR |
|---------|-------------|-----|
| `rename-selector-items-to-options` | `XDSSelector`: `items` → `options` | #479 |
| `unify-visibility-to-onOpenChange` | Renames `onHide`, `onClose`, `onToggle`, `onMenuToggle`, `onShow` → `onOpenChange` across 9 components | #473 |
| `unify-uncontrolled-to-defaultX` | `initialIsSideNavCollapsed` → `defaultIsSideNavCollapsed`, `initialIsOpen` → `defaultIsOpen`, `initialIsExpanded` → `defaultIsExpanded` | #470 |
| `rename-banner-endButton-to-endContent` | `XDSBanner`: `endButton` → `endContent` | #437 |
| `rename-form-tooltip-startIcon` | Form fields: `tooltip` → `labelTooltip`, `startIcon` → `labelIcon` (7 components) | #375 |
| `rename-isShown-to-isOpen` | `XDSDialog`, `XDSCommandPalette`, `XDSPopover`: `isShown` → `isOpen` | #472 |

### Visibility codemod details

The `unify-visibility-to-onOpenChange` transform handles the most complex case:

- **Simple renames** (1:1): `XDSDialog.onHide`, `XDSDialogHeader.onHide`, `XDSCommandPalette.onHide`, `XDSMobileNav.onClose`, `XDSDropdownMenu.onMenuToggle`, `XDSPopover.onToggle`, `XDSCollapsibleGroup.onToggle`
- **Merge handling** (XDSHoverCard, XDSTooltip): If only `onShow` or only `onHide` exists, renames to `onOpenChange`. If both exist, adds a TODO comment and leaves them for manual resolution.

## Agent docs update

Added `npx xds upgrade` and `npx xds upgrade --apply` to the compressed index injected into AGENTS.md/CLAUDE.md, with a RULE:

> After bumping @xds/core version, ALWAYS run npx xds upgrade --apply to migrate breaking API changes.

## Architecture

```
packages/cli/src/
├── commands/
│   └── upgrade.mjs              # CLI command (Commander)
└── codemods/
    ├── registry.mjs              # Version → transforms mapping
    ├── runner.mjs                # Orchestration: glob, run, report
    ├── ensure-jscodeshift.mjs    # Lazy-install on first use
    └── transforms/
        └── v0.0.2/
            ├── index.mjs         # Manifest
            ├── 6 transform files
            └── __tests__/
                └── 6 test files (61 tests total)
```

## Adding future codemods

To add codemods for a new version:

1. Create `transforms/v{X.Y.Z}/` with transform files and an `index.mjs` manifest
2. Add the version to `registry.mjs`
3. The upgrade command automatically picks up new versions

## Tests

61 tests covering all 6 transforms — positive cases, negative cases (non-XDS components), edge cases, and undefined-return-on-no-change verification.

Closes #484

---
*Crafted with care by Navi*

### CODE REVIEW

File: packages/cli/src/codemods/registry.mjs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+const registry = new Map([
+  ['0.0.2', () => import('./transforms/v0.0.2/index.mjs')],
+]);
+
+/**
+ * All registered versions, sorted ascending.
+ */
+export const versions = [...registry.keys()].sort();
```
Comment (thedjpetersen) [inline packages/cli/src/codemods/registry.mjs:15]: String comparison: `versions.filter((v) => v > from && v <= to)` will mis-order once you hit `0.0.10` — `"0.0.10" > "0.0.9"` is false in JS string ordering. Consider a numeric semver compare, or `a.localeCompare(b, undefined, {numeric: true})`.

Same pattern in `upgrade.mjs` line 97: `currentVersion >= targetVersion`.

*[Navi for DJ]*

**Review by thedjpetersen — COMMENTED** [review]
Clean architecture — the registry/runner/transform split is easy to extend. Two things I noticed:

The version comparisons in `registry.mjs` and `upgrade.mjs` use string ordering (`v > from`, `currentVersion >= targetVersion`), which works for single-digit segments but breaks at `0.0.10` — string `"0.0.10" > "0.0.2"` is false in JS. A simple `localeCompare` with `{numeric: true}` or splitting on dots would fix it before it bites you.

The agent-docs RULE for always running `xds upgrade --apply` after bumping versions is a nice touch — that will save a lot of "why did my code break" debugging.

*[Navi for DJ]*

File: packages/cli/src/codemods/registry.mjs (hunk 1 above)
Comment (thedjpetersen) [inline packages/cli/src/codemods/registry.mjs:15]: String comparison breaks at `0.0.10`: `"0.0.10" > "0.0.9"` is false in JS. Same issue in `upgrade.mjs` where `currentVersion >= targetVersion` uses string ordering. Worth adding a numeric semver compare helper before the registry grows.

*[Navi for DJ]*

**Review by thedjpetersen — COMMENTED** [review]
Clean architecture — the registry/runner/transform split is easy to extend. Two things I noticed:

The version comparisons in `registry.mjs` and `upgrade.mjs` use string ordering (`v > from`, `currentVersion >= targetVersion`), which works for single-digit segments but breaks at `0.0.10` — string `"0.0.10" > "0.0.2"` is false in JS. A simple numeric semver compare would fix this before it bites you:

```js
function semverLt(a, b) {
  const pa = a.split('.').map(Number);
  const pb = b.split('.').map(Number);
  for (let i = 0; i < 3; i++) {
    if (pa[i] !== pb[i]) return pa[i] < pb[i];
  }
  return false;
}
```

The agent-docs RULE for always running `xds upgrade --apply` after bumping versions is a nice touch.

*[Navi for DJ]*

**Review by thedjpetersen — COMMENTED** [review]
Clean architecture — the registry/runner/transform split is easy to extend. Two things I noticed:

The version comparisons in `registry.mjs` and `upgrade.mjs` use string ordering (`v > from`, `currentVersion >= targetVersion`), which works for single-digit segments but breaks at `0.0.10` — string `"0.0.10" > "0.0.2"` is false. A simple `localeCompare` with `{numeric: true}` or splitting on dots would fix it before it bites you.

The agent-docs RULE for always running `xds upgrade --apply` after bumping versions is a nice touch — that will save a lot of "why did my code break" debugging.

*[Navi for DJ]*
